fcrepo-clients/src/test/java/edu/unc/lib/dl/event/RepositoryPremisLoggerIT.java

```diff
@@ -166,4 +181,221 @@ public void getEventsModelForObjectWithoutLog() throws Exception {
+        Model initialLogModel = retrieveLogger.getEventsModel();
+        assertFalse("New premis already contains events", initialLogModel.listObjects().hasNext());
+
+        // add an event
+        Resource event = logger.buildEvent(Premis.note)
+                .addEventDetail("first premis event")
+                .write();
+
+        // create a read lock
+        PID logPid = DatastreamPids.getMdEventsPid(parentObject.getPid());
+        Lock logLock = lockManager.awaitReadLock(logPid);
+
+        // start a thread to read the log
+        List<String> premisNotes = new ArrayList<>();
+        Runnable readThreadRunnable = new Runnable() {
+            @Override
+            public void run() {
+                Model logModel = retrieveLogger.getEventsModel();
+                Resource logEventResc = logModel.getResource(event.getURI());
+                premisNotes.add(logEventResc.getProperty(Premis.note).getString());
+            }
+        };
+        Thread readThread = new Thread(readThreadRunnable);
+        readThread.start();
+        readThread.join();
+        assertEquals("first premis event", premisNotes.get(0));
+
+        // release read lock
+        logLock.unlock();
+    }
+
+    @Test
+    public void readLockWaitsForWriteLock() throws InterruptedException {
+        parentObject = repoObjFactory.createDepositRecord(null);
+        initPremisLogger(parentObject);
+
+        // make sure that there are no events in premis log before the writes
+        PremisLogger retrieveLogger = new RepositoryPremisLogger(parentObject, mockSession,
+                pidMinter, repoObjLoader, repoObjFactory);
+        Model initialLogModel = retrieveLogger.getEventsModel();
+        assertFalse("New premis already contains events", initialLogModel.listObjects().hasNext());
+
+        // add event
+        Resource event = logger.buildEvent(Premis.note)
+                .addEventDetail("first premis event")
+                .write();
+
+        // start write thread
+        Resource anotherEvent = logger.buildEvent(Premis.note)
+                .addEventDetail("second premis event")
+                .create();
+        Runnable writeThreadRunnable = new Runnable() {
+            @Override
+            public void run() {
+                logger.writeEvents(anotherEvent);
+            }
+        };
+        Thread writeThread = new Thread(writeThreadRunnable);
+        writeThread.start();
+
+        // start read thread
+        List<String> premisNotes = new ArrayList<>();
+        Runnable readThreadRunnable = new Runnable() {
+            @Override
+            public void run() {
+                Model logModel = retrieveLogger.getEventsModel();
+                Resource logEventResc = logModel.getResource(anotherEvent.getURI());
+                premisNotes.add(logEventResc.getProperty(Premis.note).getString());
+            }
+        };
+        Thread readThread = new Thread(readThreadRunnable);
```

Comment:
i'm not sure this is showing that the read waited for the write, since we aren't verifying the read hasn't completed before the write

I think you would need to start the read and write threads at the same time, so set up threads in advance and then `writeThread.start();` immediately followed by `readThread.start();`. You'd want to show that the read thread had not completed yet with `premisNotes.size() == 0` before the `writeThread.join();`, and then that the event was present after. 

In order to get the timing to work out reliably, in the writeThread you could do:
1. get a write lock for the object
2. start the writeEvents operation
3. sleep for 25ms
4. release the write lock

Reply:
When running this test locally I was curious to see what order threads were starting and stopping, so I added System.out.printlns at the beginning and end of each thread. The tiny extra delay seems to cause the readThread to perform its read before the write lock goes into place most of the time, so I get the error below, and the test fails. Did you run into this at all or on travis? I'm just a bit worried the test will flap from time to time.

```
Starting write
Starting read
Exception in thread "Thread-6" java.lang.NullPointerException
	at edu.unc.lib.dl.event.RepositoryPremisLoggerIT$5.run(RepositoryPremisLoggerIT.java:320)
	at java.lang.Thread.run(Thread.java:821)
Finished write
```

what I was running
```
        Runnable writeThreadRunnable = new Runnable() {
            @Override
            public void run() {
                System.out.println("Starting write");
                logger.writeEvents(anotherEvent);
                System.out.println("Finished write");
            }
        };
        Thread writeThread = new Thread(writeThreadRunnable);

        // start read thread
        List<String> premisNotes = new ArrayList<>();
        Runnable readThreadRunnable = new Runnable() {
            @Override
            public void run() {
                System.out.println("Starting read");
                Model logModel = retrieveLogger.getEventsModel();
                Resource logEventResc = logModel.getResource(anotherEvent.getURI());
                premisNotes.add(logEventResc.getProperty(Premis.note).getString());
                System.out.println("Finished read");
            }
        };
```

Reply:
I forgot to add the sleep between starting the write and read threads.  It is no longer flapping locally.

Reply:
hmm, they are no longer happening concurrently are they? I'd guess the write thread finishes before read starts, and the 0 size check passes because its reading the list faster than the readThread reads from the logger. I guess I was thinking the threads should start at the same time, but there would be a sleep within the write thread instead:
```
Runnable writeThreadRunnable = new Runnable() {
            @Override
            public void run() {
                ReadWriteLock writeLock = pidLockManager.awaitWriteLock();
                logger.writeEvents(anotherEvent);
                Thread.sleep(25);
                writeLock.unlock();
            }
        };
```